src/services/StreamPeriodsService.ts

```diff
@@ -0,0 +1,190 @@
+				fromUnixTime(startTimestamp),
+				fromUnixTime(endTimestamp),
+				period,
+				tokenPriceData?.prices || [],
+			),
+		};
+	});
+}
+
+/**
+ * Recursive function to virtualize stream period in different time spans. (Daily, Weekly, Monthly etc.)
+ * Amounts are calculated based on the overlapping period of date filter, stream period and virtualized period.
+ */
+function virtualizeStreamPeriod(
+	streamPeriod: StreamPeriodResult,
+	startDate: Date,
+	endDate: Date,
+	period: UnitOfTime,
+	priceData: TimespanPrice[],
+) {
+	const { flowRate, startedAtTimestamp, stoppedAtTimestamp } = streamPeriod;
+
+	const streamStoppedTimestamp = stoppedAtTimestamp || getUnixTime(Date.now());
+	const endTimestamp = Math.min(getUnixTime(endDate), streamStoppedTimestamp);
+
+	// Date when stream period and virtual period start overlapping
+	const streamPeriodStartTimestamp = Math.max(getUnixTime(startDate), startedAtTimestamp);
+
+	// Virtual period start and end timestamps (start and end timestamps of day, week, month or year)
+	const virtualPeriodEndTimestamp = getEndOfPeriodTimestamp(streamPeriodStartTimestamp, period);
+
+	// Timestamp when stream period, virtual period end and end date filter stop overlapping
+	const streamPeriodEndTimestamp = Math.min(virtualPeriodEndTimestamp, endTimestamp);
+
+	const virtualStreamPeriod: VirtualStreamPeriod = {
+		startTime: streamPeriodStartTimestamp,
+		endTime: streamPeriodEndTimestamp,
+		amount: getAmountInTimespan(streamPeriodStartTimestamp, streamPeriodEndTimestamp, flowRate).toString(),
+		amountFiat: calculateVirtualStreamPeriodPrice(
+			streamPeriodStartTimestamp,
+			streamPeriodEndTimestamp,
+			flowRate,
+			priceData,
+		).toString(),
+	};
+
+	if (endTimestamp <= virtualPeriodEndTimestamp) return [virtualStreamPeriod];
+
+	const nextPeriodStartDate = fromUnixTime(virtualPeriodEndTimestamp + 1);
+	return [
+		virtualStreamPeriod,
+		...virtualizeStreamPeriod(streamPeriod, nextPeriodStartDate, endDate, period, priceData),
+	];
+}
+
+function getAmountInTimespan(startTimestamp: number, endTimestamp: number, flowRate: string): Decimal {
+	return new Decimal(flowRate).mul(new Decimal(endTimestamp - startTimestamp));
+}
+
+function getUniqueNetworkTokenAddresses(streamPeriods: StreamPeriodResult[]): NetworkToken[] {
+	return Object.values(
+		streamPeriods.reduce((tokens, streamPeriod) => {
+			const {
+				chainId,
+				token: { id, underlyingAddress },
+			} = streamPeriod;
+
+			// TODO: If no underlyingAddress then it is native super token?
+			return {
+				...tokens,
+				[`${chainId}-${underlyingAddress}`]: { chainId, token: underlyingAddress || id },
+			};
+		}, {}),
+	);
+}
+
+function getPeriodRelevantPriceData(startTimestamp: number, endTimestamp: number, priceData: TimespanPrice[]) {
+	const priceWhenPeriodStarts = priceData
+		.filter((timespanPrice) => timespanPrice.start <= startTimestamp)
+		.reduce((startTimespanPrice: TimespanPrice | undefined, timespanPrice: TimespanPrice) => {

```

Comment:
If I understand this correctly then it would be more readable to first filter and then call some `max` function on `.start`.

Reply:
Good point. I replaced with lodash.maxBy.